Adding Yul and Yul+ to smart contract languages (#1573)
ethereum/ethereum-org-website: src/content/developers/docs/smart-contracts/languages/index.md

This example should give you a sense of what Vyper contract syntax is like. For a more detailed description of the functions and variables, [see the docs](https://vyper.readthedocs.io/en/latest/vyper-by-example.html#simple-open-auction).

## Yul and Yul+ {#yul}

> 

Reviewer: I think it might be nice to have a short introduction to these two languages. I realise that there's an intro right at the top of the page but I feel like a lot of people would miss this.

Author: Yeah but then we basically need to repeat the same all over again --> Yul is an intermediate language for Ethereum. There's not much more to say than the stuff that is mentioned in the bullet points below.

Reviewer: Why would someone want to use Yul or Yul+ ? After a Google, I came across this description of Yul+ "It allows the developer to get much closer to raw EVM than Solidity, and with that comes the promise to drastically improve gas usage."

Could we say something like "Yul gives you more control for optimising your smart contract, like reducing the amount of Gas it needs and..."

What are the other outcomes a developer might be able to achieve by using Yul instead of Solidity – hope this makes sense.

Author: It does make sense but then we're repeating what is said in the comparison section. :) There is also no such "arguments" in the Vyper or Solidity sections (just to be consistent).

Author: Also, the "Intro to Yul+" post that you googled is already linked below in the important resources section. :)

Author: I was trying to follow the structure of the page and put advantages of the language in the comparison section and the objective traits of the lang in the introductory bullet points, just like it's being done with the other two languages.

**Yul**
- Intermediate language for Ethereum.

> - Supports EVM and eWASM, and it is designed to be a usable common denominator of both platforms.

Reviewer: Can we add an explainer of what eWASM refers to and link to EVM page?

> - Support for EVM 1.0, EVM 1.5 and eWASM is planned, and it is designed to be a usable common denominator of all three platforms.

Reviewer 2: ```suggestion
- Supports EVM and eWASM, and it is designed to be a usable common denominator of both platforms.
```

> - Good target for high-level optimisation stages that can benefit all target platforms equally.

Reviewer: I'm not following this but I am also not a developer. Will defer to @samajammin here, but we may want to explain what this means a little more.

Reviewer 3: I'm also unclear on this @franzihei - mostly around what "target" is referring to.
- Does "Good target for high-level optimisation stages" == "Useful language for high-level optimisation stages"?
- Does "all target platforms"== "EVM and eWASM"?

**Yul+**
- A low-level, highly efficient extension to Yul.

> - Initially designed for an optimistic rollup contract.

Reviewer: Can we link to layer 2 scaling page here: `/en/developers/docs/layer-2-scaling/`

- Yul+ can be looked at as an experimental upgrade proposal to Yul, adding new features to it.

### Important links {#important-links-2}
- [Yul Documentation](https://solidity.readthedocs.io/en/latest/yul.html)
- [Yul+ Documentation](https://github.com/fuellabs/yulp)
- [Yul+ Playground](https://yulp.fuel.sh/)
- [Yul+ Introduction Post](https://medium.com/@fuellabs/introducing-yul-a-new-low-level-language-for-ethereum-aa64ce89512f)

> ### Example contract {#example-contract-2}

Reviewer: Excuse my ignorance but what does this contract do? We should maybe explain what this simple example does.

Author: It's just a super easy power on/off contract that doesn't "do" anything useful, just shows the language. For the more useful stuff the ERC20 contract is interesting, but very long. I don't think we need to explain it in more detail here since the people that Yul is interesting to _should_ be able to get what they need from it. Maybe we should just move the "disclaimer" that I put in the comparison section right on top of the Yul/Yul+ section so that it becomes clear that this is not necessarily for beginners? --> **If you're new to Ethereum and you haven't done any coding with smart contract languages yet, we recommend getting started with Solidity or Vyper and, only once you're familiar with smart contract security best practices and the specifics of working with the EVM, looking into Yul or Yul+.**

> The following simple example can be compiled using ``solc --strict-assembly``. 

Reviewer 2: Maybe add `--bin` because otherwise it will not output anything. Or provide a full commandline:
``solc --strict-assembly --bin input.yul``
and meniton taht the example should be stored in the file `input.yul`

```

> 

Reviewer 2: ```suggestion
```

{
    function power(base, exponent) -> result
    {
        switch exponent
        case 0 { result := 1 }
        case 1 { result := base }
        default
        {
            result := power(mul(base, base), div(exponent, 2))
            switch mod(exponent, 2)
                case 1 { result := mul(base, result) }
        }
    }

> }

Reviewer 2: I would add:
```
let result := power(calldataload(0), calldataload(32))
mstore(0, result)
return(0, 32)
```
otherwise the optimizer will completely remove the code.

Author: add it in which line?

Reviewer 2: Above the last `}`

```

> If you are already well experienced with smart contracts, a full ERC20 implementation in Yul can be found [here](https://solidity.readthedocs.io/en/v0.7.2/yul.html#complete-erc20-example).

Reviewer 2: maybe also link to `latest`?

- Good target for high-level optimisation stages that can benefit all target platforms equally.

**Yul+**
- A low-level, highly efficient extension to Yul.
- Initially designed for an [optimistic rollup](en/developers/docs/layer-2-scaling/#rollups-and-sidechains) contract.
- Yul+ can be looked at as an experimental upgrade proposal to Yul, adding new features to it.

### Important links {#important-links-2}
- [Yul Documentation](https://solidity.readthedocs.io/en/latest/yul.html)
- [Yul+ Documentation](https://github.com/fuellabs/yulp)
- [Yul+ Playground](https://yulp.fuel.sh/)
- [Yul+ Introduction Post](https://medium.com/@fuellabs/introducing-yul-a-new-low-level-language-for-ethereum-aa64ce89512f)

### Example contract {#example-contract-2}

The following simple example can be compiled using ``solc --strict-assembly --bin input.yul``. The example should
be stored in the input.yul file. 

```
{
    function power(base, exponent) -> result
    {
        switch exponent
        case 0 { result := 1 }
        case 1 { result := base }
        default
        {
            result := power(mul(base, base), div(exponent, 2))

>             switch mod(exponent, 2)
>                 case 1 { result := mul(base, result) }

Reviewer 2: Maybe we should use if now (didn't exist when the example was written):
```suggestion
            if mod(exponent, 2) { result := mul(base, result) }
```

This example should give you a sense of what Vyper contract syntax is like. For a more detailed description of the functions and variables, [see the docs](https://vyper.readthedocs.io/en/latest/vyper-by-example.html#simple-open-auction).

## Yul and Yul+ {#yul}

If you're new to Ethereum and haven't done any coding with smart contract languages yet, we recommend getting started with Solidity or Vyper and, only once you're familiar with smart contract security best practices and the specifics of working with the EVM, looking into Yul or Yul+.

**Yul**
- Intermediate language for Ethereum.
- Supports the [EVM](en/developers/docs/evm) and [eWASM](https://github.com/ewasm), an Ethereum flavored WebAssembly, and is designed to be a usable common denominator of both platforms.

> - Good target for high-level optimisation stages that can benefit all target platforms equally.

Reviewer 3: ```suggestion
- Good target for high-level optimisation stages that can benefit both EVM and eWASM platforms equally.
```

**Yul+**
- A low-level, highly efficient extension to Yul.
- Initially designed for an [optimistic rollup](en/developers/docs/layer-2-scaling/#rollups-and-sidechains) contract.
- Yul+ can be looked at as an experimental upgrade proposal to Yul, adding new features to it.

### Important links {#important-links-2}
- [Yul Documentation](https://solidity.readthedocs.io/en/latest/yul.html)
- [Yul+ Documentation](https://github.com/fuellabs/yulp)
- [Yul+ Playground](https://yulp.fuel.sh/)
- [Yul+ Introduction Post](https://medium.com/@fuellabs/introducing-yul-a-new-low-level-language-for-ethereum-aa64ce89512f)

### Example contract {#example-contract-2}

> The following simple example can be compiled using ``solc --strict-assembly --bin input.yul``. The example should

Reviewer 3: ```suggestion
The following simple example implements a power function. It can be compiled using ``solc --strict-assembly --bin input.yul``. The example should
```

### What is great about Solidity? 
- If you are a beginner, there are many tutorials and learning tools out there. See more about that in the [Learn by Coding](https://ethereum.org/en/developers/learning-tools/) section.
- Good developer tooling available.
- Solidity has a big developer community, which means you'll most likely find answers to your questions quite quickly.

### What is great about Vyper?
- Great way to get started for Python devs that want to write smart contracts.
- Vyper has a smaller number of features which makes it great for quick prototyping of ideas.
- Vyper aims to be easy to audit and maximally human-readable.

### What is great about Yul and Yul+?

> - Simplistic and functional low-level grammar. 

Reviewer 2: ```suggestion
- Simplistic and functional low-level language. 
```
why grammar?
